teps/0028-task-results-without-results.md

```diff
@@ -0,0 +1,358 @@
+skip the task.
+
+Pros/Cons:
+
++ This works well without having task results default. The `param` in the task has full control over where its value
+coming from and what is the default if it's not initialized.
++ Works with `continueAfterSkip` feature.
+- Cannot be extended to `when` expressions since task result in `when` expression is not mapped to any `param` in the
+pipeline.
+
+### Option D
+
+No defaults allowed (neither in the `task` producing results nor in the `pipelineTask` for the `task` producing results),
+just skip `finally` task with a reason when a task result is missing.
+
+Pros/Cons:
+
+- Cannot extend this option to `continueAfterSkip` with missing task results.
++ `pipeline` does not have to aware of any defaults or declare any defaults.
+- Change in perception/contract that a `finally` task is being `skipped` where it's supposed to always run. Counter
+argument to this con is, a task result is considered as a variable to a task and if the variable is not initialized,
+`finally` clause throws error and is not executed.
+
+### Option E
+
+`Pipeline` author can check the status of the result producing task and design the `pipeline` to handle success and
+failure scenarios.
+
+Pros/Cons:
+
++ `Pipeline` author has all the freedom to design the `pipeline`.
++ Can be extended to `continueAfterSkip` with missing task results (provided execution status is available within `dag`
+tasks)
+- Cannot work with `when` expressions
+- If `pipeline` author does not design failure scenario, what value should be used for the `param` referring to task
+results?
+
+### Option F
+
+The `cleanup` is necessary when the resources are acquired but not needed if the acquisition fails. This option is
+proposing introducing `when` expression to `finally`. `Cleanup` task in `finally` checks the execution status of
+`acquire` task, and guards `cleanup` if `acquire` failed. This proposal helps avoid dealing with missing task results
+in `finally` i.e. no default or no zero values for task results needed.
+
+Pros/Cons:
+
++ Introducing guarded finally task, its a clean design.
++ `Pipeline` author has full control over its design.
++ Can be extended to `continueAfterSkip`
+- no defaults needed
+- Cannot work with `when` expressions
+
+**Analysis:** We have listed `when` expression with missing task results as one of the possible use case and something
+needs to addressed with the proposal in this TEP. But having difficulty coming up with a possibility of getting into
+`when` expression with missing task results. Let's take an example for the same,
+A(guarded) -> B(continueAfterSkip=true)(guarded with task results from A)  -> C
+This example shows that `B` is guarded with task results from `A` where `A` was skipped but is it really a possible
+scenario? When this `pipeline` says continue executing `B` even though `A` was skipped, why would `B` be guarded with a
+task result from `A`?
+
```

Comment:
another alternative would be to provide/require a completely different way of accessing results for the use cases above

i.e. if you want to look at a result value, and you want to run regardless of the state of the producing task (i.e. even if failed, skipped, etc.), we provide another way to provide this info, e.g.:

1. write a file to disk with the status of the entire pipelinerun
2. provide variable replacement for the entire pipelinerun status

i think that providing something like (1) might be a reasonable alternative here; i have a feeling that the tasks that we want to create for the use case above might be more complex than just the boskos example we are including

Reply:
@chhsia0 also mentioned another interesting option in the working group today which i think it worth exploring: introducing a syntax for the variable replacement that allows a default to be specified, e.g.:

`$(tasks.boskos-acquire.results.leased-resource | fake-awesome-project)`

I really like that this option would let us both:
1. Let the consumer of the value decide how to handle the case where it is not provided
2. Work for when expressions AND params

However if we go that route i think we'd also want to take into account #1393 i.e. what would this syntax look like in a world where we have more types than just strings and can do stuff like array indexing

Reply:
I also liked the idea of expressing a default value in the variable substitution syntax itself, especially if it looks somewhat similar to Bash syntax, e.g. `$(tasks.boskos-acquire.results.leased-resource:-fake-awesome-project}`.  This is a well-known feature of shell scripting so people would just "get it" without a lot of explanation required.  I guess the downside is that you need to specify the default on every reference as opposed to having a single place where you could set it as in Priti's proposal.

Reply:
> another alternative would be to move the "default" specification into the "params" section in the pipelineTask instead

Yup, this can also allow specifying separate defaults for each reference, but downside with `when expressions` is significant as well. How would we allow having defaults with `when expressions` then?

Reply:
thank you for the suggestion, have included as one of alternatives

Reply:
I liked the proposed options. But also defaulting on the producer and consumer side do not have to be orthogonal. 
Defaulting a result would be an helper for task author to ensure a result is always provided, that the author has enough context in the task to be able to do that meaningfully.
Defaulting a variable value would be a helper the task author (or runner) to provide a default value for situations where the pipeline / pipelinerun context is required to provide a meaningful default.
Both could be implemented as separate features.